lib/services/access_request.go

```diff
@@ -902,12 +902,61 @@ Outer:
 	return len(needsAllow) == 0, nil
 }
 
-func NewReviewPermissionChecker(ctx context.Context, getter RequestValidatorGetter, username string) (ReviewPermissionChecker, error) {
+type userStateRoleOverride struct {
```

Comment:
I'm not too sure about how "weird" this wrapper is. It felt weirder to introduce a `SetRoles` method on `UserState` since this was primarily read-only.

Reply:
I see value in keeping `UserState` read-only, so this is fine by me.  I'd suggest adding a simple test to make sure that the roles set in the override end up actually being the ones used by the permission checker (just to make sure that if we started using a different method for loading roles in the future, we don't accidentally start ignoring the override).